src/cursor/abstract_cursor.ts

```diff
@@ -691,243 +695,156 @@ export abstract class AbstractCursor<
-    }
-
-    if (cursor[kDocuments].length === 0 && blocking === false) {
-      if (!shift) return false;
-      return null;
-    }
-  } while (!cursor.isDead || cursor[kDocuments].length !== 0);
-
-  if (!shift) return false;
-  return null;
-}
-
-async function cleanupCursor(
-  cursor: AbstractCursor,
-  options: { error?: AnyError | undefined; needsToEmitClosed?: boolean } | undefined
-): Promise<void> {
-  const cursorId = cursor[kId];
-  const cursorNs = cursor[kNamespace];
-  const server = cursor[kServer];
-  const session = cursor[kSession];
-  const error = options?.error;
-
-  // Cursors only emit closed events once the client-side cursor has been exhausted fully or there
-  // was an error.  Notably, when the server returns a cursor id of 0 and a non-empty batch, we
-  // cleanup the cursor but don't emit a `close` event.
-  const needsToEmitClosed = options?.needsToEmitClosed ?? cursor[kDocuments].length === 0;
-
-  if (error) {
-    if (cursor.loadBalanced && error instanceof MongoNetworkError) {
-      return await completeCleanup();
+      await this.cleanup();
     }
   }
 
-  if (cursorId == null || server == null || cursorId.isZero() || cursorNs == null) {
-    if (needsToEmitClosed) {
-      cursor[kClosed] = true;
-      cursor[kId] = Long.ZERO;
-      cursor.emit(AbstractCursor.CLOSE);
-    }
-
-    if (session) {
-      if (session.owner === cursor) {
+  /** @internal */
+  private async cleanup(error?: Error) {
+    this[kClosed] = true;
+    const session = this[kSession];
+    try {
+      if (
+        !this[kKilled] &&
+        this[kId] &&
+        !this[kId].isZero() &&
+        this[kNamespace] &&
+        this[kServer] &&
+        !session.hasEnded
+      ) {
+        this[kKilled] = true;
+        await executeOperation(
+          this[kClient],
+          new KillCursorsOperation(this[kId], this[kNamespace], this[kServer], { session })
+        );
+      }
+    } catch (error) {
+      squashError(error);
+    } finally {
+      if (!session.hasEnded && session?.owner === this) {
         await session.endSession({ error });
-        return;
       }
-
-      if (!session.inTransaction()) {
+      if (!session?.inTransaction()) {
         maybeClearPinnedConnection(session, { error });
       }
-    }
 
-    return;
+      this.emitClose();
+    }
   }
 
-  async function completeCleanup() {
-    if (session) {
-      if (session.owner === cursor) {
-        try {
-          await session.endSession({ error });
-        } finally {
-          cursor.emit(AbstractCursor.CLOSE);
-        }
-        return;
-      }
-
-      if (!session.inTransaction()) {
-        maybeClearPinnedConnection(session, { error });
+  /** @internal */
+  private hasEmittedClose = false;
+  /** @internal */
+  private emitClose() {
+    try {
+      if (!this.hasEmittedClose && (this[kDocuments].length === 0 || this[kClosed])) {
+        // @ts-expect-error: not sure..
+        this.emit('close');
```

Comment:
If we're going to leave a ts-expect-error in, can we document why it's an error?  but tbh I'd expect this to work, not sure why an expect error is needed here.

Reply:
Yes sorry meant to get back to this is is declared in the same way ConnectionEvents are so I'm not sure what is going on here.

```ts
Argument of type '[]' is not assignable to parameter of type 'Parameters<CursorEvents["close"]>'.ts(2345)
```

Maybe it is because CursorEvents is generic so Parameters<CursorEvents["close"]> may not be assignable to `[]`. Not sure how to require extenders do not add parameters.